File: 2023/04/solution.exs
defmodule Part1 do
  defp solve([], sum), do: sum

  defp solve([{winning, mine} | cards], sum) do
    overlap = MapSet.intersection(winning, mine) |> MapSet.size()

    if overlap > 0 do
      solve(cards, sum + 2 ** (overlap - 1))
    else
      solve(cards, sum)
    end
  end

  def solve(cards) do
    solve(cards, 0)
  end
end

defmodule Part2 do
  defp solve([], _, _, sum), do: Map.values(sum) |> Enum.sum()

  defp solve([{winning, mine} | cards], i, len, sum) do
    overlap = MapSet.intersection(winning, mine) |> MapSet.size()
    sum = Map.update(sum, i, 1, &(&1 + 1))

    sum =
      if overlap > 0 do
        0..(overlap - 1)
        |> Enum.map(fn j -> i + 1 + j end)
        |> Enum.filter(fn j -> j < len end)
        |> Enum.reduce(sum, fn j, acc ->
          curr = Map.get(acc, i, 0)
          Map.update(acc, j, curr, &(&1 + curr))
        end)
      else
        sum
      end

    solve(cards, i + 1, len, sum)
  end

  def solve(lines), do: solve(lines, 0, Enum.count(lines), Map.new())
end

defmodule Parse do
  def parse_numbers(line) do
    line
    |> String.trim()
    |> String.split(~r/\s+/)
    |> Enum.map(&String.to_integer/1)
  end
end

# Read all lines from stdin
lines =
  IO.read(:stdio, :eof)
  |> String.split("\n")
  |> Enum.map(fn line ->
    [left, my_cards] = String.split(line, " | ")
    [_game_id, winning_cards] = String.split(left, ": ")

    my_cards = Parse.parse_numbers(my_cards)
    winning_cards = Parse.parse_numbers(winning_cards)

    {MapSet.new(winning_cards), MapSet.new(my_cards)}
  end)

IO.write("Part 1: ")
Part1.solve(lines) |> IO.inspect()

IO.write("Part 2: ")
Part2.solve(lines) |> IO.inspect()

File: 2023/04/solution.ml
let lines =
  Seq.of_dispenser (fun _ ->
    match read_line () with
    | x -> Some x
    | exception End_of_file -> None)
  |> List.of_seq

module IntSet = Set.Make (struct
    type t = int

    let compare = compare
  end)

let parse_card str =
  let [ left; right ] = String.split_on_char '|' str in
  let [ _; winning ] = String.split_on_char ':' left in
  let winning =
    Str.split (Str.regexp " +") winning |> List.map int_of_string |> IntSet.of_list
  in
  let mine =
    Str.split (Str.regexp " +") right |> List.map int_of_string |> IntSet.of_list
  in
  (winning, mine)
[@@warning "-8"]

let cards = List.map parse_card lines

let score = function
  | 0 -> 0
  (* lsl = left shift logical, doubling via bit hackery *)
  | n -> 1 lsl (n - 1)

let play (winning, mine) = IntSet.inter winning mine |> IntSet.to_seq |> Seq.length

let part1 () =
  cards
  |> List.map play
  |> List.map score
  |> List.fold_left ( + ) 0
  |> Printf.printf "Part 1: %d\n"

let play_part2 cards_held card_number cards =
  let current_count = Hashtbl.find cards_held card_number in
  let score = play cards in
  Seq.ints 1
  |> Seq.take score
  |> Seq.map (fun x -> x + card_number)
  |> Seq.filter (fun x -> Hashtbl.mem cards_held x)
  |> Seq.iter (fun x ->
    let curr = Hashtbl.find cards_held x in
    Hashtbl.replace cards_held x (curr + current_count))

let calc_score cards = Hashtbl.fold (fun _ v acc -> acc + v) cards 0

let part2 () =
  let len = List.length cards in
  let cards_held =
    Seq.ints 0 |> Seq.take len |> Seq.map (fun x -> (x, 1)) |> Hashtbl.of_seq
  in
  cards
  |> List.to_seq
  |> Seq.mapi (fun i (winning, mine) -> (i, (winning, mine)))
  |> Seq.iter (fun (i, cards) -> play_part2 cards_held i cards);
  calc_score cards_held |> Printf.printf "Part 2: %d\n"

let () =
  part1 ();
  part2 ()
